### ISSUE TITLE
Wrong AST max complexity computation

### ISSUE THREAD

**Issue by N0xFF** [rmosolgo/graphql-ruby#2419]
To reproduce an error
```ruby
# spec/support/dummy/schema.rb

class MaxSchema < GraphQL::Schema
  query DairyAppQuery

  max_depth 3
  max_complexity 1

  use GraphQL::Execution::Interpreter
  use GraphQL::Analysis::AST
end
```
```ruby
# spec/graphql/analysis/ast/max_query_complexity_spec.rb

describe "should not affected by other plugins" do
  let(:schema) { Dummy::MaxSchema }
  let(:query_string) {%|
    {
      a: cheese(id: 1) { ...cheeseFields }
      b: cheese(id: 1) { ...cheeseFields }
      c: cheese(id: 1) { ...cheeseFields }
      d: cheese(id: 1) { ...cheeseFields }
      e: cheese(id: 1) { ...cheeseFields }
    }

    fragment cheeseFields on Cheese { id }
  |}
  let(:result) { schema.execute(query_string) }

  it "returns an error" do
    assert_equal "Query has complexity of 10, which exceeds max complexity of 1", result["errors"].first["message"]
  end
end
```

**Expected**

`Query has complexity of 10, which exceeds max complexity of 1`

**Actual**

`Query has complexity of 15, which exceeds max complexity of 1`

Problem appeared after https://github.com/rmosolgo/graphql-ruby/pull/2389, but probably it's result of a more complex problem. In quick look, I noticed more frequent call `AST::QueryComplexity#on_leave_field` after adding `max_depth` in schema.

**Comment by rmosolgo** [comment]
Thanks for reporting this issue! 🤔 I wonder how that snuck in...

**Comment by jturkel** [comment]
Doh. It looks like both [QueryComplexity#on_enter_fragment_spread](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/analysis/ast/query_complexity.rb#L54) and [QueryDepth#on_enter_fragment_spread](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/analysis/ast/query_depth.rb#L63) call ` visitor.enter_fragment_spread_inline(node)` with will force the fragment nodes to be visited for *all* analyzers. Prior to #2389 this didn't matter since we were always maintaining complexities in a hash so the duplicates were ignored because the response keys were the same.  #2389 optimized away this hash for certain cases so we're now double counting. How hard would it be to modify the visitor framework so visiting fragment inline spreads didn't affect other visitors?

**Comment by swalkinshaw** [comment]
The real fix here is to properly support visiting fragment definitions without having to force inline visits.

I've run into the same issue for other analyzers and talked to @xuorig about it but I don't think it's too easy to solve (from memory). I believe this is the root cause: https://github.com/swalkinshaw/graphql-ruby/blob/61a0669e7e92a9a5a5ec6b1d18d7fbf104f8d3b4/lib/graphql/analysis/ast/visitor.rb#L26

Since it only visits the selected operation (which we want), it also skips fragment *definitions*.

**Comment by swalkinshaw** [comment]
One idea I had was to use the existing `DefinitionSlice` to  build a new document from the operation definition we want + its dependencies (fragment definitions).

**Comment by jturkel** [comment]
This is probably a silly question but when would analyzers *not* want fragment definitions to be automatically inlined by the framework? Out of the three analyzers that ship with the gem both `GraphQL::Analysis::AST::Complexity` and `GraphQL::Analysis::AST::QueryDepth` perform the inlining and `GraphQL::Analysis::AST::FieldUsage` doesn't but should.

**Comment by xuorig** [comment]
@jturkel not a silly question at all. I think @rmosolgo and I initially wanted to keep it pretty "raw" / bare bones, analyzing the unchanged AST. If I remember well we had in mind to probably eventually create an implementation of it that is more user friendly / more specific on top of that barebone analyzer class. Maybe it's a good time to start working on it 🤔

**Comment by swalkinshaw** [comment]
I forgot to mention that part... it should be the default I agree! I remember talking to @rmosolgo or @xuorig (maybe both) and we couldn't really think of use cases against it.

**Comment by jturkel** [comment]
I took a crack at a fix in #2421. Feedback would be welcomed!

**Comment by rmosolgo** [comment]
> use the existing `DefinitionSlice` to build a new document

🤔 Wow, that's really intriguing. It would simplify the visit a lot, although visitors would have to do more work to piece the spreads and fragments back together. And I've never investigated that operation from a performance perspective, so it would probably benefit from a quick look. 

> fragment definitions to be automatically inlined

That's a great suggestion. `InternalRepresentation::Rewrite` basically _did_ that (and a lot of other things that were slow), and it made analyzers pretty easy. 

I wonder if there's something to this: make a pass of the AST to _prepare_ it for analysis, inlining fragments. Then visit the prepared document which is already inlined. But we'd have to retain references to the original AST nodes for error reporting.
